Search for the prerequisites from package configuration file.

blattms opened the pull request:
That is the way it is done in other modules, too. We need this
at least for libecl as this target is used in opm-common_LIBRARIES
and not defined if there was no find_package(ecl) call. That call
is now made during execution of find_package_deps(opm-common).

This is an alternative to #457 and prevents users from manually searching for libecl.

akva2 commented on opm-common-prereqs.cmake:
```diff
@@ -25,3 +25,5 @@ else()
         "Boost 1.44.0
           COMPONENTS system unit_test_framework REQUIRED")
 endif()
+# We the a defined target for libecl
```
wat?

blattms replied:
thanks. Improved

akva2 commented:
i prefer this but then we can remove the other lib handling for ecl in OpmFind, no ?

blattms commented:
Will give that a try and report back.

blattms commented:
@akva2 you mean all of [OpmFind.cmake#L165-L173](/OPM/opm-common/blob/19afcbf769e8c4e11316248be49156ea2dd4b43d/cmake/Modules/OpmFind.cmake#L165-L173)?

akva2 commented:
i mean the lib part. some stuff still might want the header paths.

blattms commented:
to the very least we need to have `set(ECL_LIBRARIES ecl)` as libecl does not export the target name of the library.

akva2 commented:
yes that is what i meant. library as ecl, only dig out headers in case something wants those.

blattms commented:
Seems like we still need these. If I just use `set(ecl_LIBRARIES ecl)` then opm-common does not compile:
```
[ 19%] Building CXX object CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o
/usr/bin/c++   -DHAVE_CONFIG_H=1 -I/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel -I/home/mblatt/src/dune/opm-2.6/opm-common/external/cjson -I/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel/include -I/home/mblatt/src/dune/opm-2.6/opm-common  -std=c++14 -pipe -Wall -Wno-unknown-pragmas -pthread -O3 -mtune=native -DNDEBUG -fPIC   -std=c++14 -o CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o -c /home/mblatt/src/dune/opm-2.6/opm-common/src/opm/parser/eclipse/Deck/Deck.cpp
In file included from /home/mblatt/src/dune/opm-2.6/opm-common/opm/parser/eclipse/Deck/Deck.hpp:32:0,
                 from /home/mblatt/src/dune/opm-2.6/opm-common/src/opm/parser/eclipse/Deck/Deck.cpp:23:
/home/mblatt/src/dune/opm-2.6/opm-common/opm/parser/eclipse/Units/UnitSystem.hpp:28:30: fatal error: ert/ecl/ecl_util.h: No such file or directory
 #include <ert/ecl/ecl_util.h>
                              ^
compilation terminated.
CMakeFiles/opmcommon.dir/build.make:873: recipe for target 'CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o' failed
make[2]: *** [CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o] Error 1
make[2]: Leaving directory '/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel'
CMakeFiles/Makefile2:1321: recipe for target 'CMakeFiles/opmcommon.dir/all' failed
make[1]: *** [CMakeFiles/opmcommon.dir/all] Error 2
make[1]: Leaving directory '/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

blattms commented:
Nevermind my last comment. I had a typo in the variable name ecl_LIBRARIES. will retest

akva2 commented:
nice.

bska commented on opm-common-prereqs.cmake:
```diff
@@ -25,3 +25,7 @@ else()
+# We need a defined target for libecl when linking.
+# The definition of the target is done implicitly below when
+# libecl is searched for.
+find_package_deps(opm-common)
```
Do you really mean `opm-common` here (in `opm-common-prereqs.cmake`)?  Is `opm-common` a prerequisite for itself?

akva2 replied:
it's the prefix for the variables used inside the macros. opm-common_CONFIG_VAR, opm-common_DEPS.

bska replied:
> it's the prefix for the variables used inside the macros

Ah, right.  Thanks for clarifying and sorry for the confusion.

blattms commented:
Fixed one commit message with a forced push. Sorry.

I have tested this branch (cherry-picked to releases/2018.04) with @joakim-hove's described in #456 and it worked on my system with extra patches to sunbeam. Additionally I tried it together with opm-grid and dumux. In contrast to @andlaus I can build dumux without problems (by passing -DUSE_MPI=ON to cmake to activate MPI for OPM). Either his setup is very different (which I cannot judge because his report is too detailed and might well be) or I fixed some bugs in this branch without really noticing.

blattms commented:
jenkins build this please

blattms commented:
BTW we are still only check ENABLE_ECL_INPUT and neglect ENABLE_ECL_OUTPUT in the package configuration file. Is that really correct. i.e. does ENABLE_ECL_OUTPUT being ON imply the same for ENABLE_ECL_OUTPUT?

akva2 commented:
OUTPUT implies INPUT, https://github.com/OPM/opm-common/blob/master/CMakeLists.txt#L12

both are exported through the config file, https://github.com/OPM/opm-common/blob/master/CMakeLists.txt#L73

joakim-hove commented:
jenkins build this please

akva2 commented:
this will still break due to bugs in the boost find module.

akva2 commented:
might have found a local issue on jenkins. let's try again and see if it fixed anything.

akva2 commented:
jenkins build this please

blattms commented:
When I looked at the output, I saw a linking error in upscaling (boost iostream). Will test this locally this afternoon.

akva2 commented:
i cannot trigger it locally, only on jenkins, which is why i'm looking into local issues.

blattms commented:
I have to explicitly use `DINSTALL_BENCHMARKS=ON` with cmake to reproduce the problem on my machine. Otherwise upscale_relperm_benchmark will not be built.

blattms commented:
`-lboost_iostream` seems to missing from the linker line. I guess that problem is in opm-upscaling?

blattms commented:
Well, somehow searching for the same lib multiple times must have been lost. Ergo we neglect the components requested in opm-upscaling. Will fix.

blattms commented:
jenkins build this please

blattms commented:
This should be fixed now. 

Indeed the reason for this failure was that we only searched for Boost once after my changes. The reason was that I stripped the variable opm_common_processed that basically made us search for everything required above of the line for opm-common in opm-module-prereq.cmake multiple times (including Boost by accident or by design). With my changes these multiple searches were removed. With my last commit another search is now always triggered if the argument list to find_package contains the word COMPONENTS.
